<%inherit file="/sequence/view_base.html" />
<%namespace file="/components/tw_bootstrap.html" name="boot" import="*" />
<%namespace file="/sequence/partials.html" name="part" import="*" />

% if not c.sequence_group.analyzed:
	% if c.pending_job:
		${part.job_css()}
	<div class="well">
		<p>
		The genomic sequences and SNPs near this assay are being analyzed.  This typically takes from 30 seconds to a few minutes, depending on current traffic levels.  If you feel an excessive amount of time has passed, or the job has been aborted, email <a href="mailto:jeff.mellen@quantalife.com">jeff.mellen@quantalife.com</a> for further support.
		</p>
		<h2>Processing Status</h2>
		${part.job_details(c.pending_job)}
	</div>
	% else:
	<div class="well">
		The server was unable to retrieve sequence and SNP data for this assay.  Please contact <a href="mailto:jeff.mellen@quantalife.com">jeff.mellen@quantalife.com</a> for further assistance.assistance.
	</div>
	% endif
% elif not (c.sequences or c.transcripts):
	<div class="well">
		<div style="text-align: center">
			% if c.display_mode == 'transcript':
				No transcripts found.
			% else:
				No amplicons found.
			% endif
		</div>
	</div>
% endif
<% idx = 0 %>
% if c.sequences:
% for seq in c.sequences:
	<% idx = idx+1 %>
	<h2>Match #${idx}:</h2>
	${part.seq_result(seq, verified=c.primer_amplicons)}
	% if hasattr(seq, 'snps') and seq.snps:
		${part.snp_results(seq.snps)}
	% endif
% endfor
% elif c.transcripts:
% for genomic_region, transcripts in c.transcripts:
	<% idx = idx+1 %>
	<h2>Match #${idx}:</h2>
	${part.transcript_result(genomic_region, transcripts, verified=c.primer_amplicons)}
	% if hasattr(transcripts[0], 'snps') and transcripts[0].snps:
		${part.snp_results(transcripts[0].snps)}
	% endif
% endfor
% endif
